contributors/design-proposals/daemonset-create-first-rolling-update.md

```diff
@@ -0,0 +1,82 @@
+## DaemonSet create-first rolling update strategy
+
+DaemonSets currently support a "delete first, then add" rolling update
+strategy.
+
+This document presents a design for a new "add first, then delete" rolling
+update strategy, called `SurgingRollingUpdate`. This strategy will create new
+pods first, cleaning up the old pods once the new ones are running. As with
+Deployments, users will use a `maxSurge` parameter to control how many nodes
+are allowed to have >1 pod running during a rolling update.
+
+### Example use case
+
+For self-hosted Kubernetes DaemonSets are a natural fit for running the
+scheduler and controller-manager on master nodes. However, since DaemonSets are
+managed by the controller-manager, if `maxUnavailable` >= the number of masters
+then a rolling update of the controller-manager would delete all the old pods
+before the new pods were scheduled, leading to a control plane outage.
+
+This feature is a blocker for supporting self-hosted Kubernetes in `kubeadm`.
+
+### Implementation plan
+
+A new `SurgingRollingUpdate` strategy will be added to the `DaemonSet` API
+object. Then an additional hook for this strategy will be added to the part of
+the reconciliation loop that runs when pod updates are needed. That hook will:
+
+1. Determine how many new pods can be created, according to the number of extra
+   pods currently running and the `maxSurge` parameter.
+2. Determine how many previous-generation pods can be retired once their
+   replacement pods are scheduled and running.
+3. Instruct the controller-manager to add and delete pods based on the results
+   of (1) and (2).
+
+This is mechanically almost identical to the current `RollingUpdate` strategy,
+with the key differences being the direct creation of new pods (the
+`RollingUpdate` strategy only deletes pods up to the `maxUnavailable` threshold
+and then waits for the main reconciler create new pods to replace them).
+
+One minor complication is that the controller-manager's reconciliation loop
+actively enforces the one-pod-per-node invariant. This invariant will need to
+be relaxed when the `SurgingRollingUpdate` strategy temporarily creates extra
+pods.
+
+### Alternatives considered
+
+The `maxSurge` parameter could have been added to the existing `RollingUpdate`
+strategy (as is the case for Deployments). However, this would break backwards
+compatibility, since the `maxUnavailable` parameter currently requires
+`maxUnavailable` to be > 0, but this is not required with
+`SurgingRollingUpdate`.
+
+There are plans to create a `v1beta2` iteration of the extensions API, so it
+may be possible to add the `maxSurge` strategy to `RollingUpdate` and deprecate
+`SurgingRollingUpdate` then.
+
+For more background see https://github.com/kubernetes/kubernetes/issues/48841.
+
+### Considerations / questions
+
+1. How are `hostPort`s handled? 
```

Comment:
aren't hostPorts likely to be common in self-hosted scenarios (one of the driving use cases for this proposal)?

Reply:
SO_REUSEPORT ?

What I alluded to with "really clever work" would be to emulate that in the port-forwarding  path, but that sounds disastrous.

Reply:
I would argue that special cases for hostPorts, hostNetworking, and other scarce resources like GPUs should not be a goal here. In particular, the behavior you describe is also true of Deployments.

For example (and I tested this), create a deployment where replicas = worker nodes, and use a hostport. Set `maxUnavailable` to 0, and then update your deployment. The new pods will be stuck in pending since the scheduler can't place the new pods anywhere.

This behavior is also true for DaemonSets as you point out (updated in the doc). And yes, I can see how DaemonSets may more often use hostPorts / hostNetworking in practice, but I think the scheduler's job is to place pods subject to the intersection of their resource constraints, not do resource planning for you.

We should absolutely call out this risk in the documentation. But I don't think that it should be this code's job to try to foresee potential resource conflicts.

Reply:
I basically agree.  I don't think it is tractable to try to fix this special case.

Reply:
This should be documented and highlighted.m  The documentation can point the user to use the "delete before add" model. 

Also, it is possible to test for a pod having a hostPath and refusing to do a "add before delete" option? (updated: it is noted in the document on line 136)